File: csharp/CsSearch/CsSearchLib/Searcher.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

using CsFindLib;

namespace CsSearchLib;

public class Searcher
{
	private readonly Finder _finder;
	private SearchSettings Settings { get; }
	private ConcurrentBag<SearchResult> Results { get; }
	private Encoding TextFileEncoding { get; set; } = Encoding.Default;

	// TODO: move this to SearchSettings
	private const int FileBatchSize = 255;
	// use single-byte encoding for reading binary files (will break if UTF8)
	private readonly Encoding _binaryEncoding = Encoding.GetEncoding("ISO-8859-1");

	public Searcher(SearchSettings settings)
	{
		Settings = settings;
		try
		{
			_finder = new Finder(settings);
		}
		catch (FindException e)
		{
			throw new SearchException(e.Message);
		}
		ValidateSettings();
		Results = [];
	}

	private void ValidateSettings()
	{
		if (Settings.SearchPatterns.Count < 1)
			throw new SearchException("No search patterns defined");
		try
		{
			TextFileEncoding = Encoding.GetEncoding(Settings.TextFileEncoding);
		}
		catch (ArgumentException)
		{
			throw new SearchException("Invalid encoding");
		}
		if (Settings.LinesBefore < 0)
			throw new SearchException("Invalid linesbefore");
		if (Settings.LinesAfter < 0)
			throw new SearchException("Invalid linesafter");
		if (Settings.MaxLineLength < 0)
			throw new SearchException("Invalid maxlinelength");
	}

	private void AddSearchResult(SearchResult searchResult)
	{
		Results.Add(searchResult);
	}

	private static bool LinesMatch(IEnumerable<string> lines,
		ICollection<Regex> inPatterns, ICollection<Regex> outPatterns)
	{
		var lineList = lines.ToList();
		return ((inPatterns.Count == 0 || AnyMatchesAnyPattern(lineList, inPatterns))
		        && (outPatterns.Count == 0 || !AnyMatchesAnyPattern(lineList, outPatterns)));
	}

	private bool LinesBeforeMatch(IEnumerable<string> linesBefore)
	{
		return LinesMatch(linesBefore, Settings.InLinesBeforePatterns,
			Settings.OutLinesBeforePatterns);
	}

	private bool LinesAfterMatch(IEnumerable<string> linesAfter)
	{
		return LinesMatch(linesAfter, Settings.InLinesAfterPatterns,
			Settings.OutLinesAfterPatterns);
	}

	private static IEnumerable<int> GetNewLineIndices(string text)
	{
		// to keep it simple, only get indices of '\n' and deal with '\r' later
		IList<int> newLineIndices = new List<int>();
		for (var i = 0; i < text.Length; i++)
		{
			if (text[i] == '\n')
				newLineIndices.Add(i);
		}
		return newLineIndices;
	}

	private static IEnumerable<int> GetStartLineIndices(IEnumerable<int> newLineIndices)
	{
		var startLineIndices = new List<int> {0};
		startLineIndices.AddRange(newLineIndices.Select(i => i + 1));
		return startLineIndices;
	}

	private static IEnumerable<int> GetEndLineIndices(string text,
		IEnumerable<int> newLineIndices)
	{
		var endLineIndices = new List<int>(newLineIndices) {text.Length - 1};
		return endLineIndices;
	}

	private IEnumerable<string> GetLinesBeforeFromContents(string content,
		IEnumerable<int> beforeStartIndices, IEnumerable<int> beforeEndIndices)
	{
		var linesBefore = new List<string>();
		if (Settings.LinesBefore == 0) return linesBefore;
		var starts = beforeStartIndices.Reverse().Take(Settings.LinesBefore).Reverse().ToList();
		var ends = beforeEndIndices.Reverse().Take(Settings.LinesBefore + 1).Reverse().Skip(1).ToList();
		linesBefore.AddRange(starts.Select(
			(t, i) => content.Substring(t, ends[i] - t)));
		return linesBefore;
	}

	private IEnumerable<string> GetLinesAfterFromContents(string content,
		IEnumerable<int> afterStartIndices, IEnumerable<int> afterEndIndices)
	{
		var linesAfter = new List<string>();
		if (Settings.LinesAfter == 0) return linesAfter;
		var starts = afterStartIndices.Take(Settings.LinesAfter).ToList();
		var ends = afterEndIndices.Skip(1).Take(Settings.LinesAfter).ToList();
		linesAfter.AddRange(starts.Select(
			(t, i) => content.Substring(t, ends[i] - t)));
		return linesAfter;
	}

	public IEnumerable<SearchResult> SearchContents(string contents)
	{
		var patternMatches = new Dictionary<Regex, int>();
		var results = new List<SearchResult>();
		var newLineIndices = GetNewLineIndices(contents).ToList();
		var startLineIndices = GetStartLineIndices(newLineIndices).ToList();
		var endLineIndices = GetEndLineIndices(contents, newLineIndices).ToList();

		foreach (var p in Settings.SearchPatterns)
		{
			var match = p.Match(contents);
			while (match.Success)
			{
				if (Settings.FirstMatch && patternMatches.ContainsKey(p))
				{
					break;
				}
				var matchIndex = match.Index;
				// get the start and end indices before the match index
				var beforeStartIndices = startLineIndices.TakeWhile(i => i <= matchIndex).ToList();
				var beforeEndIndices = endLineIndices.TakeWhile(i => i < matchIndex).ToList();
				// add another end line index if it exists or the last index of the string
				if (endLineIndices.Count > beforeEndIndices.Count)
					beforeEndIndices.ToList().Add(endLineIndices[beforeEndIndices.Count]);
				else
					beforeEndIndices.ToList().Add(contents.Length - 1);
				var afterStartIndices = startLineIndices.SkipWhile(i => i <= matchIndex).ToList();
				var afterEndIndices = endLineIndices.SkipWhile(i => i <= matchIndex).ToList();
				var lineNum = beforeStartIndices.ToList().Count;
				var startLineIndex = beforeStartIndices.Max();
				var endLineIndex = afterStartIndices.Min() - 1;
				var line = contents.Substring(startLineIndex, endLineIndex - startLineIndex);
				var linesBefore = GetLinesBeforeFromContents(contents,
					beforeStartIndices.AsEnumerable().Reverse().Skip(1).Reverse(),
					beforeEndIndices).ToList();
				var linesAfter = GetLinesAfterFromContents(contents, afterStartIndices,
					afterEndIndices).ToList();
				if ((linesBefore.ToList().Count == 0 || LinesBeforeMatch(linesBefore))
				    &&
				    (linesAfter.ToList().Count == 0 || LinesAfterMatch(linesAfter)))
					results.Add(new SearchResult(
						p,
						null,
						lineNum,
						match.Index - startLineIndex + 1,
						match.Index - startLineIndex + match.Length + 1,
						line,
						new List<string>(linesBefore),
						new List<string>(linesAfter)));
				patternMatches[p] = 1;

				match = match.NextMatch();
			}
		}
		return results;
	}

	private static bool AnyMatchesAnyPattern(IEnumerable<string> strings,
		IEnumerable<Regex> patterns)
	{
		return strings.Any(s => MatchesAnyPattern(s, patterns));
	}

	private static bool MatchesAnyPattern(string s, IEnumerable<Regex> patterns)
	{
		return !string.IsNullOrEmpty(s) && patterns.Any(p => p.Match(s).Success);
	}

	public IEnumerable<SearchResult> SearchLines(IEnumerable<string> lines)
	{
		var patternMatches = new Dictionary<Regex, int>();
		var results = new List<SearchResult>();
		var lineNum = 0;
		var linesBefore = new Queue<string>();
		var linesAfter = new Queue<string>();

		using var lineEnumerator = lines.GetEnumerator();
		while (linesAfter.Count > 0 || lineEnumerator.MoveNext())
		{
			lineNum++;
			var line = linesAfter.Count > 0 ? linesAfter.Dequeue() : lineEnumerator.Current;
			if (Settings.LinesAfter > 0)
			{
				while (linesAfter.Count < Settings.LinesAfter && lineEnumerator.MoveNext())
				{
					linesAfter.Enqueue(lineEnumerator.Current);
				}
			}

			if ((Settings.LinesBefore == 0 || linesBefore.Count == 0 || LinesBeforeMatch(linesBefore))
			    &&
			    (Settings.LinesAfter == 0 || linesAfter.Count == 0 || LinesAfterMatch(linesAfter)))
			{
				foreach (var p in Settings.SearchPatterns)
				{
					var matches = new List<Match>();
					if (Settings.FirstMatch)
					{
						var match = p.Match(line);
						if (match.Success)
						{
							matches.Add(match);
							patternMatches[p] = 1;
						}
					}
					else
					{
						var mc = p.Matches(line);
						if (mc.Count > 0)
						{
							matches.AddRange(mc);
							patternMatches[p] = 1;
						}
					}
					foreach (var match in matches)
					{
						results.Add(new SearchResult(p,
							null,
							lineNum,
							match.Index + 1,
							match.Index + match.Length + 1,
							line,
							new List<string>(linesBefore),
							new List<string>(linesAfter)));
					}
				}
			}

			// If all search patterns are in patternMatches, FirstMatch complete, return results
			if (Settings.FirstMatch && patternMatches.Count == Settings.SearchPatterns.Count)
			{
				return results;
			}

			if (Settings.LinesBefore == 0) continue;
			if (linesBefore.Count == Settings.LinesBefore)
			{
				linesBefore.Dequeue();
			}
			if (linesBefore.Count < Settings.LinesBefore)
			{
				linesBefore.Enqueue(line);
			}
		}

		return results;
	}

	private void SearchTextFileLines(FileResult f)
	{
		try
		{
			var enumerableLines = File.ReadLines(f.FullName);
			var results = SearchLines(enumerableLines);

			foreach (var r in results)
			{
				r.File = f;
				AddSearchResult(r);
			}
		}
		catch (IOException e)
		{
			Logger.Log(e.Message);
		}
	}

	private void SearchTextFileContents(FileResult f)
	{
		try
		{
			var contents = FileUtil.GetFileContents(f.FullName, TextFileEncoding);
			var results = SearchContents(contents);
			foreach (var r in results)
			{
				r.File = f;
				AddSearchResult(r);
			}
		}
		catch (IOException e)
		{
			Logger.Log(e.Message);
		}
	}

	private void SearchTextFile(FileResult f)
	{
		if (Settings.Debug)
			Logger.Log($"Searching text file {f.PathAndName}");
		if (Settings.MultiLineSearch)
			SearchTextFileContents(f);
		else
			SearchTextFileLines(f);
	}

	private void SearchBinaryFile(FileResult fr)
	{
		if (Settings.Verbose)
			Logger.Log($"Searching binary file {fr.PathAndName}");
		try
		{
			using var sr = new StreamReader(fr.FullName, _binaryEncoding);
			var contents = sr.ReadToEnd();
			foreach (var p in Settings.SearchPatterns)
			{
				if (Settings.FirstMatch)
				{
					var m = p.Match(contents);
					if (m.Success)
					{
						AddSearchResult(new SearchResult(
							p,
							fr,
							0,
							m.Index + 1,
							m.Index + m.Length + 1,
							null));
					}
				}
				else
				{
					var matches = p.Matches(contents).Cast<Match>();
					foreach (var m in matches)
					{
						AddSearchResult(new SearchResult(
							p,
							fr,
							0,
							m.Index + 1,
							m.Index + m.Length + 1,
							null));
					}
				}
			}
		}
		catch (IOException e)
		{
			Logger.Log(e.Message);
		}
	}

	public void SearchFile(FileResult f)
	{
		switch (f.Type)
		{
			case FileType.Code:
			case FileType.Text:
			case FileType.Xml:
				SearchTextFile(f);
				break;
			case FileType.Audio:
			case FileType.Binary:
			case FileType.Font:
			case FileType.Image:
			case FileType.Video:
				SearchBinaryFile(f);
				break;
			case FileType.Archive:
				Logger.Log($"Skipping archive file {f.PathAndName}");
				break;
			default:
			{
				if (Settings.Verbose)
				{
					Logger.Log($"Skipping file {f.PathAndName}");
				}

				break;
			}
		}
	}

	private void SearchBatchConcurrent(IReadOnlyList<FileResult> searchFiles)
	{
		var searchTasks = new Task[searchFiles.Count];
		for (var i = 0; i < searchFiles.Count; i++)
		{
			var searchFile = searchFiles[i];
			searchTasks[i] = Task.Factory.StartNew(() => SearchFile(searchFile));
		}
		Task.WaitAll(searchTasks);
	}

	private void SearchBatch(IReadOnlyList<FileResult> findFiles)
	{
		if (findFiles.Count > 100)
		{
			SearchBatchConcurrent(findFiles);
		}
		else
		{
			foreach (var f in findFiles)
			{
				SearchFile(f);
			}
		}
	}

	public List<SearchResult> Search()
	{
		var fileResults = _finder.Find().ToList();

		if (Settings.Verbose)
		{
			var findDirs = fileResults
				.Where(fr => fr.FilePath.Parent != null)
				.Select(fr => fr.FilePath.Parent!.ToString())
				.Distinct()
				.OrderBy(d => d).ToArray();
			Logger.Log($"Directories to be searched ({findDirs.Length}):");
			foreach (var d in findDirs)
			{
				Logger.Log(d);
			}
				
			Logger.Log($"\nFiles to be searched ({fileResults.Count}):");
			foreach (var f in fileResults)
			{
				Logger.Log(f.PathAndName);
			}
		}

		var searched = 0;
		while (fileResults.Count - searched > FileBatchSize)
		{
			SearchBatch(fileResults.Skip(searched).Take(FileBatchSize).ToArray());
			searched += FileBatchSize;
		}
		if (fileResults.Count > searched)
		{
			SearchBatch(fileResults.Skip(searched).ToArray());
		}

		var results = Results.ToList();
		SortSearchResults(results);
		return results;
	}

	private Comparison<SearchResult> GetSearchResultsComparison()
	{
		if (Settings.SortDescending)
		{
			return Settings.SortBy switch
			{
				SortBy.FileName => (r1, r2) => r2.CompareByName(r1, Settings.SortCaseInsensitive),
				SortBy.FileSize => (r1, r2) => r2.CompareBySize(r1, Settings.SortCaseInsensitive),
				SortBy.FileType => (r1, r2) => r2.CompareByType(r1, Settings.SortCaseInsensitive),
				SortBy.LastMod => (r1, r2) => r2.CompareByLastMod(r1, Settings.SortCaseInsensitive),
				_ => (r1, r2) => r2.CompareByPath(r1, Settings.SortCaseInsensitive)
			};
		}

		return Settings.SortBy switch
		{
			SortBy.FileName => (r1, r2) => r1.CompareByName(r2, Settings.SortCaseInsensitive),
			SortBy.FileSize => (r1, r2) => r1.CompareBySize(r2, Settings.SortCaseInsensitive),
			SortBy.FileType => (r1, r2) => r1.CompareByType(r2, Settings.SortCaseInsensitive),
			SortBy.LastMod => (r1, r2) => r1.CompareByLastMod(r2, Settings.SortCaseInsensitive),
			_ => (r1, r2) => r1.CompareByPath(r2, Settings.SortCaseInsensitive)
		};
	}

	private void SortSearchResults(List<SearchResult> results)
	{
		var comparison = GetSearchResultsComparison();
		results.Sort(comparison);

		if (Settings.SortDescending)
		{
			results.Reverse();
		}
	}

	public static void PrintResults(IEnumerable<SearchResult> results, SearchResultFormatter formatter)
	{
		// File sorting is done by CsFind, so maybe additional sorting isn't needed?
		// var sortedResults = GetSortedSearchResults();
		// var formatter = new SearchResultFormatter(Settings);
		var resultsList = results.ToList();
		Logger.Log($"Search results ({resultsList.Count}):");
		foreach (var searchResult in resultsList)
		{
			Logger.Log(formatter.Format(searchResult));
		}
	}

	private static IEnumerable<FileResult> GetMatchingFiles(IEnumerable<SearchResult> results)
	{
		return
		[
			..results.Where(r => r.File != null)
				.Select(r => r.File!)
				.DistinctBy(fp => fp.ToString())
		];
	}

	public static void PrintMatchingDirs(IEnumerable<SearchResult> results, SearchResultFormatter formatter)
	{
		var matchingFiles = GetMatchingFiles(results);
		Finder.PrintMatchingDirs(matchingFiles, formatter.FileFormatter);
	}

	public static void PrintMatchingFiles(IEnumerable<SearchResult> results, SearchResultFormatter formatter)
	{
		var matchingFiles = GetMatchingFiles(results);
		Finder.PrintMatchingFiles(matchingFiles, formatter.FileFormatter);
	}

	private List<string> GetMatchingLines(IEnumerable<SearchResult> results)
	{
		var lines = results.Where(r => r.Line != null)
			.Select(r => r.Line!.Trim()).ToList();
		if (Settings.UniqueLines)
		{
			lines = new HashSet<string>(lines).ToList();
		}
		lines.Sort(new CaseInsensitiveComparer());
		return lines;
	}

	public void PrintMatchingLines(IEnumerable<SearchResult> results, SearchResultFormatter formatter)
	{
		var matchingLines = GetMatchingLines(results).ToList();
		var hdrText = Settings.UniqueLines ? "Unique matching lines" : "Matching lines";
		Logger.Log($"\n{hdrText} ({matchingLines.Count}):");
		foreach (var m in matchingLines)
		{
			Logger.Log(formatter.FormatLine(m));
		}
	}
}

internal class CaseInsensitiveComparer : IComparer<string>
{
	public int Compare(string? a, string? b)
	{
		return a switch
		{
			null when b == null => 0,
			null => -1,
			not null when b == null => 1,
			_ => string.Compare(a.ToUpper(), b.ToUpper(), StringComparison.Ordinal)
		};
	}
}

File: fsharp/FsSearch/FsSearchLib/Searcher.fs
namespace FsSearchLib

open System
open System.IO
open System.Text
open System.Text.RegularExpressions
open FsFindLib

type Searcher (settings : SearchSettings) =
    let _finder = Finder(settings)
    let _binaryEncoding = Encoding.GetEncoding("ISO-8859-1")
    let mutable _textFileEncoding = Encoding.GetEncoding("utf-8")

    // read-only member properties
    member this.Finder = _finder
    member this.BinaryEncoding = _binaryEncoding

    // read-write member properties
    member this.TextFileEncoding 
        with get () = _textFileEncoding
        and set enc = _textFileEncoding <- enc

    // member methods
    member this.EncodingFromStringOrError (encName : string) : string option =
        try
            this.TextFileEncoding <- Encoding.GetEncoding(encName)
            None
        with
        | :? ArgumentException ->
            Some $"Invalid encoding: %s{encName}"

    member this.ValidateSettings () : string list =
        match _finder.ValidateSettings() with
        | [] ->
            [
                (if List.isEmpty settings.SearchPatterns then (Some "No search patterns defined") else None);
                (if settings.LinesBefore < 0 then (Some "Invalid linesbefore") else None);
                (if settings.LinesAfter < 0 then (Some "Invalid linesafter") else None);
                (if settings.MaxLineLength < 0 then (Some "Invalid maxlinelength") else None);
                (this.EncodingFromStringOrError settings.TextFileEncoding);
            ]
            |> List.choose id
        | errs -> errs

    member this.MatchesAnyPattern (s : string) (patterns : Regex list) : bool =
        Seq.exists (fun p -> (p:Regex).Match(s).Success) patterns

    member this.AnyMatchesAnyPattern (slist : string seq) (patterns : Regex list) : bool =
        Seq.exists (fun s -> this.MatchesAnyPattern s patterns) slist

    member this.recGetPatternMatches (patterns : Regex list) (s : string) (pMatches : (Regex * Match) list) = 
        match patterns with
        | [] -> pMatches
        | p :: ps ->
            let allMatches = [for m in p.Matches(s) do yield m]
            let nextMatches =
                if allMatches.IsEmpty
                then []
                else
                    if settings.FirstMatch
                    then allMatches |> List.take 1
                    else allMatches
            let nextPMatches = List.map (fun m -> (p, m)) nextMatches
            this.recGetPatternMatches ps s (List.append pMatches nextPMatches)

    member this.GetLineIndices (s : string) : (int * int) list =
        let newLineIndices =
            Array.zip [|0 .. s.Length - 1|] (s.ToCharArray())
            |> Array.filter (fun x -> snd x = '\n')
            |> Array.map (fun x -> fst x)
            |> List.ofArray
        let startLineIndices =
             0 :: List.map (fun i -> i + 1) newLineIndices
             |> Seq.takeWhile (fun i -> i < s.Length)
             |> List.ofSeq
        //Logger.Log (sprintf "startLineIndices (%d): [%s]" startLineIndices.Length (String.Join(",", startLineIndices)))
        let endLineIndices = newLineIndices
        //Logger.Log (sprintf "endLineIndices (%d): [%s]" endLineIndices.Length (String.Join(",", endLineIndices)))
        if endLineIndices.Length < startLineIndices.Length then
            List.zip startLineIndices (endLineIndices @ [s.Length - 1])
        else
            List.zip startLineIndices endLineIndices

    member this.LinesMatch (lines : string seq, inPatterns : Regex list,
                            outPatterns : Regex list) : bool =
        (List.isEmpty inPatterns || (this.AnyMatchesAnyPattern lines inPatterns))
        &&
        (List.isEmpty outPatterns || (not (this.AnyMatchesAnyPattern lines outPatterns)))

    member this.LinesBeforeMatch (linesBefore : string seq) : bool =
        this.LinesMatch (linesBefore, settings.InLinesBeforePatterns,
                         settings.OutLinesBeforePatterns)

    member this.LinesAfterMatch (linesAfter : string seq) : bool =
        this.LinesMatch(linesAfter, settings.InLinesAfterPatterns,
                        settings.OutLinesAfterPatterns)

    member this.TakeRight (ss : 'T list) (num : int) : 'T list =
        if ss.Length < num then
            ss
        else
            ss
            |> List.ofSeq
            |> List.rev
            |> Seq.take num
            |> List.ofSeq
            |> List.rev

    member this.SearchContents (s : string) : SearchResult.t list =
        let mutable results : SearchResult.t list = []
        let lineIndices = this.GetLineIndices s
        for p in settings.SearchPatterns do
            let mutable m = p.Match s
            let mutable stop = false
            while m.Success && not stop do
                let matchStartIndex = m.Index
                let matchEndIndex = m.Index + m.Length
                let startLineIndex = Seq.filter (fun i -> fst i < matchStartIndex) lineIndices |> Seq.max |> fst
                //Logger.Log (sprintf "startLineIndex: %d" startLineIndex)
                let endLineIndex = Seq.filter (fun i -> snd i > matchStartIndex) lineIndices |> Seq.min |> snd
                //Logger.Log (sprintf "endLineIndex: %d" endLineIndex)
                let beforeIndices = Seq.takeWhile (fun i -> fst i < startLineIndex) lineIndices |> List.ofSeq
                //Logger.Log (Common.ListToString("beforeIndices", beforeIndices))
                let afterIndices = Seq.skipWhile (fun i -> fst i <= matchStartIndex) lineIndices |> List.ofSeq
                let lineNum = beforeIndices.Length + 1
                let line = s.Substring(startLineIndex, endLineIndex - startLineIndex)
                let beforeLineIndices = this.TakeRight beforeIndices settings.LinesBefore
                //Logger.Log (Common.ListToString("beforeLineIndices", beforeLineIndices))
                let linesBefore =
                    if settings.LinesBefore > 0 then
                        beforeLineIndices
                        |> List.map (fun (x,y) -> s.Substring(x, y - x))
                    else
                        []
                let linesAfter =
                    if settings.LinesAfter > 0 then
                        afterIndices
                        |> Seq.take(settings.LinesAfter)
                        |> List.ofSeq
                        |> List.map (fun (x,y) -> s.Substring(x, y - x))
                    else
                        []
                if (linesBefore.Length = 0 || this.LinesBeforeMatch linesBefore) &&
                   (linesAfter.Length = 0 || this.LinesAfterMatch linesAfter) then
                    let r = SearchResult.Create p lineNum (matchStartIndex - startLineIndex + 1) (matchEndIndex - startLineIndex + 1) line linesBefore linesAfter
                    results <- List.append results [r]
                    if settings.FirstMatch then
                        stop <- true
                m <- m.NextMatch()
        results

    member this.SearchTextFileContents (f : FileResult.t) : SearchResult.t list =
        let mutable results : SearchResult.t list = []
        try
            let contents = FileUtil.GetFileContents f.File.FullName this.TextFileEncoding
            results <-
                this.SearchContents contents
                |> List.map (fun r -> { r with File = f })
        with
        | :? IOException as e -> printfn $"%s{e.Message}"
        results

    member this.SearchLine (line : string) (lineNum : int) (linesBefore : string list) (linesAfter : string list) : SearchResult.t list =
        let pMatches = this.recGetPatternMatches settings.SearchPatterns line []
        let toSearchResult (p : Regex) (m : Match) = 
            SearchResult.Create p lineNum (m.Index + 1) (m.Index + m.Length + 1) line linesBefore linesAfter
        let results : SearchResult.t list = List.map (fun (p, m) -> toSearchResult p m) pMatches
        results

    member this.SearchLines (lines : string list) =
        let rec recSearchLines (lineNum : int) (linesBefore : string list) (lines : string list) (results : SearchResult.t list) =
            let nextLines =
                if settings.FirstMatch && (List.length results) = (List.length settings.SearchPatterns)
                then []
                else lines
            match nextLines with
            | [] -> results
            | l :: ls ->
                let linesAfter =
                    if settings.LinesAfter > 0
                    then
                        if (List.length ls < settings.LinesAfter)
                        then ls
                        else (List.take settings.LinesAfter ls)
                    else []
                let rs = this.SearchLine l lineNum linesBefore linesAfter
                let lbs = 
                    if settings.LinesBefore > 0
                    then
                        if (List.length linesBefore) = settings.LinesBefore
                        then (List.append (List.tail linesBefore) [l])
                        else (List.append linesBefore [l])
                    else []
                recSearchLines (lineNum + 1) lbs ls (List.append results rs)
        recSearchLines 1 [] lines []

    member this.SearchTextFileLines (f : FileResult.t) : SearchResult.t list =
        let mutable results : SearchResult.t list = []
        try
            let lines = File.ReadLines f.File.FullName |> List.ofSeq
            results <-
                this.SearchLines lines
                |> List.map (fun r -> { r with File = f })
        with
        | :? IOException as e -> printfn $"%s{e.Message}"
        results

    member this.SearchTextFile (f : FileResult.t) : SearchResult.t list =
        if settings.Debug then
            Logger.Log $"Searching text file %s{f.File.FullName}"
        if settings.MultiLineSearch then
            this.SearchTextFileContents f
        else
            this.SearchTextFileLines f

    member this.SearchBlob (blob : string): SearchResult.t list =
        let pMatches = this.recGetPatternMatches settings.SearchPatterns blob []
        let toSearchResult (p : Regex) (m : Match) = 
            SearchResult.Create p 0 (m.Index + 1) (m.Index + m.Length + 1) "" [] []
        let results : SearchResult.t list = List.map (fun (p, m) -> toSearchResult p m) pMatches
        results

    member this.SearchBinaryFile (f : FileResult.t) : SearchResult.t list =
        let mutable results : SearchResult.t list = []
        if settings.Verbose then
            Logger.Log $"Searching binary file %s{f.File.FullName}"
        try
            use sr = new StreamReader (f.File.FullName, this.BinaryEncoding)
            let contents = sr.ReadToEnd()
            for p in Seq.filter (fun p -> (p:Regex).Match(contents).Success) settings.SearchPatterns do
                let mutable m = p.Match contents
                let mutable stop = false
                while m.Success && not stop do
                    let matchStartIndex = m.Index + 1
                    let matchEndIndex = m.Index + m.Length + 1
                    let r = SearchResult.Create p 0 matchStartIndex matchEndIndex null [] []
                    results <- List.append results [{ r with File=f }]
                    if settings.FirstMatch then
                        stop <- true
                    m <- m.NextMatch()
        with
        | :? IOException as ex -> printfn $"%s{ex.Message}"
        results

    member this.SearchFile (f : FileResult.t) : SearchResult.t list =
        match f.FileType with
        | FileType.Archive ->
            Logger.Log "Archive file searching not currently supported"
            []
        | FileType.Binary -> this.SearchBinaryFile f
        | FileType.Code | FileType.Text | FileType.Xml -> this.SearchTextFile f
        | FileType.Unknown ->
            Logger.Log "Skipping file of unknown type"
            []
        | _ ->
            Logger.Log $"Skipping file of indeterminate type (this shouldn't happen): %s{f.File.FullName}"
            []

    member this.SortByMatchLocation (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let lineNumCmp = r1.LineNum - r2.LineNum
        if lineNumCmp = 0
        then
            let startIndexCmp = r1.MatchStartIndex - r2.MatchStartIndex
            if startIndexCmp = 0 then r1.MatchEndIndex - r2.MatchEndIndex else startIndexCmp
        else lineNumCmp

    member this.SortByPath (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let cmp = this.Finder.SortByPath r1.File r2.File
        if cmp = 0 then this.SortByMatchLocation r1 r2 else cmp

    member this.SortByName (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let cmp = this.Finder.SortByName r1.File r2.File
        if cmp = 0 then this.SortByMatchLocation r1 r2 else cmp

    member this.SortBySize (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let cmp = this.Finder.SortBySize r1.File r2.File
        if cmp = 0 then this.SortByMatchLocation r1 r2 else cmp

    member this.SortByType (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let cmp = this.Finder.SortByType r1.File r2.File
        if cmp = 0 then this.SortByMatchLocation r1 r2 else cmp

    member this.SortByLastMod (r1 : SearchResult.t) (r2 : SearchResult.t) : int =
        let cmp = this.Finder.SortByLastMod r1.File r2.File
        if cmp = 0 then this.SortByMatchLocation r1 r2 else cmp

    member this.GetSortComparator : SearchResult.t -> SearchResult.t -> int =
        if settings.SortDescending then
            match settings.SortBy with
            | SortBy.FileName -> (fun r1 r2 -> this.SortByName r2 r1)
            | SortBy.FileSize -> (fun r1 r2 -> this.SortBySize r2 r1)
            | SortBy.FileType -> (fun r1 r2 -> this.SortByType r2 r1)
            | SortBy.LastMod  -> (fun r1 r2 -> this.SortByLastMod r2 r1)
            | _               -> (fun r1 r2 -> this.SortByPath r2 r1)
        else
            match settings.SortBy with
            | SortBy.FileName -> this.SortByName
            | SortBy.FileSize -> this.SortBySize
            | SortBy.FileType -> this.SortByType
            | SortBy.LastMod  -> this.SortByLastMod
            | _               -> this.SortByPath

    member this.SortSearchResults (results : SearchResult.t list) : SearchResult.t list = 
        let sortComparator = this.GetSortComparator
        List.sortWith sortComparator results

    member this.Search () : SearchResult.t list =
        let files = _finder.Find()
        let results = files |> List.collect this.SearchFile
        this.SortSearchResults results

    member this.PrintResults (results : SearchResult.t list) (formatter : SearchResultFormatter) : unit =
        if results.Length > 0 then
            Logger.Log $"\nSearch results (%d{results.Length}):"
            results |> Seq.iter (fun r -> Logger.Log $"%s{formatter.Format r}")
        else
            Logger.Log $"\nSearch results: 0"

    member this.PrintMatchingDirs (results : SearchResult.t list) (formatter : SearchResultFormatter) : unit = 
        let files = results |> List.map _.File
        this.Finder.PrintMatchingDirs files formatter.FileFormatter

    member this.PrintMatchingFiles (results : SearchResult.t list) (formatter : SearchResultFormatter) : unit = 
        let files = results |> List.map _.File |> List.distinct
        this.Finder.PrintMatchingFiles files formatter.FileFormatter

    member this.GetMatchingLines (results : SearchResult.t list) : string list = 
        let lines =
            results
            |> Seq.map (fun r -> r.Line.Trim())
        if settings.UniqueLines then
            Seq.sortBy (fun (s : string) -> s.ToUpper()) lines
            |> Seq.distinct
            |> List.ofSeq
        else
            Seq.sortBy (fun (s : string) -> s.ToUpper()) lines
            |> List.ofSeq

    member this.PrintMatchingLines (results : SearchResult.t list) (formatter : SearchResultFormatter) : unit = 
        let lines = this.GetMatchingLines results
        let title =
            if settings.UniqueLines then "Unique matching lines"
            else "Matching lines"
        if lines.Length > 0 then
            Logger.Log $"\n%s{title} (%d{lines.Length}):"
            for l in lines do
                printfn $"%s{formatter.FormatLine(l)}"
        else
            Logger.Log $"\n%s{title}: 0"

;;
